### PR TITLE
Added keyboard movement

### PR DESCRIPTION
**Pull request by Toniman575**
Simply added constant keyboard movement to the `do_camera_movement` system and `MoveMode` and `DirectionKeys` fields to `PanCam` and registered them in Bevy.

As I do not know if this change is even desired, this seemed to be the least intrusive API/code change.
Further changes I thought about but discarded because of reason above:
- Option for accelerated movement.
- Splitting the `do_camera_movement` into multiple systems.
- Simultaneous mouse and keyboard `MoveMode`.

Addresses #52

### CODE REVIEW

**Comment by johanhelsing** [comment]
I think this makes sense scope-wise. Curious to hear anybody else's opinion as well...

API-wise: I think simultaneous movement should be implemented and be the default. Perhaps even the only mode, and remove the mode field. If you don't want keyboard controls, you can set `direction_keys = DirectionKeys::NONE`.

Curious for other thoughts on this as well.

File: src/lib.rs [hunk 1]
```diff
@@ -196,21 +219,81 @@ fn do_camera_movement(
...
+                    .down
+                    .iter()
+                    .any(|key| keyboard_buttons.pressed(*key))
+                {
+                    direction.y += 1.;
+                }
+
+                if direction.length() == 0. {
+                    continue;
+                }
```
Comment (tomara-x) [inline src/lib.rs:284]: not sure why we need to have this. i removed it and it works. or if i'm misunderstanding the point, it would be faster to use `length_squared`

File: src/lib.rs [hunk 2]
```diff
@@ -296,6 +385,14 @@ impl PanCam {
 impl Default for PanCam {
     fn default() -> Self {
         Self {
+            move_mode: MoveMode::Keyboard,
```
Comment (tomara-x) [inline src/lib.rs:388]: my other thought is: should this mode be the default? zooming is still done by scrolling, so this feels off

File: src/lib.rs (hunk 1 above)
Comment (johanhelsing) [reply src/lib.rs:284]: Struggle a bit with reading this on my phone, but I'd assume avoiding triggering change detection when there is no actual change.

Comment (Toniman575) [reply src/lib.rs:284]: Indeed that is why I put it there. I'll look into `length_squared` tomorrow.

File: src/lib.rs (hunk 2 above)
Comment (Toniman575) [reply src/lib.rs:388]: As I agree with johanhelsing that we should just implement simultaneous movement as the only mode I'll just remove `MoveMode` tomorrow when I update the pr.

**Comment by Toniman575** [comment]
The implementation I went for seemed the most sane to me but I'm curious what u guys think.
Further thoughts:
- We could move `projection.area.size()` behind if checks so it doesn't happen if unnecessary, but the performance gain seems rather insignificant and I am hesitant to add code complexity.
- I think `check_egui_wants_focus` makes enough sense as it is right now, but we could check for `wants_pointer_input` and `wants_keyboard_input` separately depending on the input of the user. The easiest way seems to be to split `do_camera_movement` system into two separate systems for keyboard movement and mouse movement. Again though, this might be a bit too intrusive of a change, especially as this change would only be relevant for a crate feature.

File: src/lib.rs [hunk 3]
```diff
@@ -196,21 +208,82 @@ fn do_camera_movement(
...
+        if cam
+            .move_keys
+            .down
+            .iter()
+            .any(|key| keyboard_buttons.pressed(*key))
+        {
+            direction.y += 1.;
+        }
+
+        let keyboard_delta = if direction.length_squared() == 0. {
```
Comment (tomara-x) [inline src/lib.rs:265]: i didn't think of this yesterday, but this is checking that the vec is zeros, right? so `if direction == Vec2::ZERO` would be more readable? unless my brain isn't working :D

i like having one mode like this. it's nice!

File: src/lib.rs [hunk 4]
```diff
@@ -222,6 +295,10 @@ fn do_camera_movement(
 #[derive(Component, Reflect)]
 #[reflect(Component)]
 pub struct PanCam {
+    /// The keyboard keys that will be used to move the camera
+    pub move_keys: DirectionKeys,
+    /// Speed for keyboard movement
```
Comment (johanhelsing) [inline src/lib.rs:300]: Maybe explain how it is proportional to the projection size?

File: src/lib.rs [hunk 5]
```diff
@@ -196,21 +208,82 @@ fn do_camera_movement(
...
+        }
+
+        if cam
+            .move_keys
+            .down
+            .iter()
+            .any(|key| keyboard_buttons.pressed(*key))
+        {
+            direction.y += 1.;
+        }
```
Comment (johanhelsing) [inline src/lib.rs:263]: Maybe move this to a method on `DirectionKeys`?

File: src/lib.rs (hunk 3 above)
Comment (johanhelsing) [reply src/lib.rs:265]: Yes, I also think a simple zero check would be more readable.

I also wonder if it's better to just let the delta variables be `Vec2` instead of `Option<Vec2>`?

and then simply do:

```rust
let delta = mouse_delta + keyboard_delta;

if delta == Vec2::ZERO {
    continue;
}
```

Would that work? Maybe my brain is also not fully operational :P

File: src/lib.rs [hunk 6]
```diff
@@ -21,13 +21,23 @@ pub struct PanCamPlugin;
...
 pub struct PanCamSystemSet;
 
+/// Which keys move the camera in particular directions for keyboard movement.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Reflect)]
+pub struct DirectionKeys {
+    up: Vec<KeyCode>,
+    down: Vec<KeyCode>,
+    left: Vec<KeyCode>,
+    right: Vec<KeyCode>,
+}
```
Comment (johanhelsing) [inline src/lib.rs:31]: Should we add some constants here for convenience? `NONE`, `ARROWS`, `WASD`, `ARROWS_AND_WASD`?

**Review by johanhelsing — CHANGES_REQUESTED** [review]
Thanks, API-wise, I think this adds it in a way that fits well with the scope and the rest of this crate.

We should update the readme/crate docs, *maybe* examples and this is good to go :)

File: src/lib.rs [hunk 7]
```diff
@@ -196,21 +208,82 @@ fn do_camera_movement(
...
+            .any(|key| keyboard_buttons.pressed(*key))
+        {
+            direction.x += 1.;
+        }
+
+        if cam
+            .move_keys
+            .right
+            .iter()
+            .any(|key| keyboard_buttons.pressed(*key))
```
Comment (johanhelsing) [inline src/lib.rs:242]: I think perhaps you could also do

```rust
keyboard_buttons.any_pressed(cam.move_keys.right)
```

Not sure which reads better?

Comment (Toniman575) [reply src/lib.rs:246]: `any_pressed` requires ownership of `KeyCode` which does not implement `copy`. I think if we can avoid cloning here we should.

File: src/lib.rs (hunk 6 above)
Comment (Toniman575) [reply src/lib.rs:35]: I had to add these with methods (other than NONE), because allocations are not allowed in constants.

**Comment by Toniman575** [comment]
I added a bit of Documentation, but I'm not sure if its too wordy so I'll defer to @johanhelsing on that. Otherwise I'm fairly happy with the changes, hope u guys are as well.

File: src/lib.rs [hunk 8]
```diff
@@ -21,13 +21,89 @@ pub struct PanCamPlugin;
...
+            left: vec![KeyCode::ArrowLeft, KeyCode::KeyA],
+            right: vec![KeyCode::ArrowRight, KeyCode::KeyD],
+        }
+    }
+
+    fn direction(&self, keyboard_buttons: &Res<ButtonInput<KeyCode>>) -> Vec2 {
+        let mut direction = Vec2::ZERO;
+
+        if self.left.iter().any(|key| keyboard_buttons.pressed(*key)) {
+            direction.x += 1.;
```
Comment (johanhelsing) [inline src/lib.rs:80]: Hmm... I think these signs should be flipped. At least that's how my brain expect it to work.

Comment (Toniman575) [reply src/lib.rs]: I amended the commit, sorry I have no idea how I managed to do that lol.

Comment (johanhelsing) [reply src/lib.rs]: Haha, yeah I was wondering if this was some kind of apple "natural" scrolling thing :P

**Review by johanhelsing — APPROVED** [review]
Great work! 

Thanks also @tomara-x for reviewing!

❤️
